include/nil/crypto3/math/polynomial/polynomial_dfs.hpp

```diff
@@ -851,29 +852,58 @@ namespace nil {
             template<typename FieldType>
             static inline polynomial_dfs<typename FieldType::value_type> polynomial_product(
                     std::vector<math::polynomial_dfs<typename FieldType::value_type>> multipliers) {
+                // Pre-create all the domains. We could do this on-the-go, but we want this function to be more
+                // parallelization-friendly. This single-threaded version may look a bit complicated,
+                // but it's now very similar to what we have in parallel code.
                 std::unordered_map<std::size_t, std::shared_ptr<evaluation_domain<FieldType>>> domain_cache;
+
+                std::size_t min_domain_size = std::numeric_limits<std::size_t>::max();
+                std::size_t total_degree = 0;
+                for (std::size_t i = 0; i < multipliers.size(); i++) {
+                    min_domain_size = std::min(min_domain_size, multipliers[i].size());
+                    total_degree += multipliers[i].degree();
```

Comment:
We need to also consider the possibility of resizes due to domain size, not just from degree sizes branch of
```
const std::size_t new_domain_size =
    detail::power_of_two(std::max(
        current_domain_size,
        next_domain_size,
        multipliers[i].degree() + multipliers[i + stride].degree() + 1}));
```

Reply:
Yes, changed.